gtk4: fix FileChooserDialog with language bindings that uses dlopen

chuangzhu opened the pull request:
###### Description of changes

https://discourse.gnome.org/t/strange-filechooserdialog-behaviour-in-gtk-4-10/14435

Fixes https://github.com/NixOS/nixpkgs/pull/226138#issuecomment-1509615369

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

jtojnar commented:
:+1:  on backporting this.

bjornfor commented on pkgs/development/libraries/gtk/4.x.nix:
```diff
@@ -81,6 +82,10 @@ stdenv.mkDerivation rec {
     # https://github.com/NixOS/nixpkgs/pull/218143#issuecomment-1501059486
     ./patches/4.0-fix-darwin-build.patch
+    (fetchpatch {
+      url = "https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/5696.patch";
```
Looks like a mutable URL, which may have different content over time.

chuangzhu committed:
gtk4: fix FileChooserDialog with language bindings that uses dlopen

https://discourse.gnome.org/t/strange-filechooserdialog-behaviour-in-gtk-4-10/14435

Fixes https://github.com/NixOS/nixpkgs/pull/226138#issuecomment-1509615369

jtojnar commented:
Looks like this is already part of [4.10.2](https://gitlab.gnome.org/GNOME/gtk/-/commits/4.10.2).

bobby285271 commented:
nautilus 44.1 now [requires](https://gitlab.gnome.org/GNOME/nautilus/-/commit/ee1d823737518a28eec51e330564cace3bf5fcd9) gtk 4.10.3, can we just bump gtk4 in master since it has 509 rebuild on x86_64-linux?

jtojnar commented:
Sounds good.
